AP_Generator: Add Intelligent Energy hydrogen fuel cell telemetry driver to AP_Generator

Gone4Dirt opened the pull request:
The primary purpose of this PR is to add telemetry drivers for three Intelligent Energy fuel cell units 650W, 800W, and 2.4kW.

These have been added into AP_Generator, in which I have created a frontend/backend for.  The previous Richenpower generator is now in the backend.

I wanted to raise this PR now to start the conversation/review process going.  I still need to adapt the auto test for the Richen Power generator to handle the new generator type parameter.

AP_generator has been moved to AP_Vehicle class to give access to Rover as well.

Two BATT<X>_MONITOR types are being used.  Allowing for 1 instance to report the electrical telemetry and the 2nd instance to report the fuel telemetry.  The electrical telemetry is largely the same as any other battery monitor.  The fuel level, piggy-backs on the consumed_mah state.  When users set BATT<X>_CAPACITY to 100 mAh then the GCS will report an easily readable % remaining.  A constant 1V is reported for the fuel instance, this is as per the FuelFlow monitor type.

I have tried to crudely show the flow of information with the following diagram:

![image](https://user-images.githubusercontent.com/34512430/96924791-12a99980-14ab-11eb-8e8a-039dfb0008f5.png)

andyp1per commented:
Richenpower has been on my hit list for something that should be off by default on 1mb boards - maybe you could make that change here since the size of the driver has increased now?

Gone4Dirt commented:
Yeah, sure thing.

Gone4Dirt commented:
@andyp1per yeah, sure thing.

peterbarker commented on libraries/AP_Generator/AP_Generator.h:
```diff
@@ -0,0 +1,110 @@
+    };
+
+    // Helper to get param and cast to GenType
+    enum GenType get_gen_type(void) const;
```
We don't usually add helpers for this sort of thing.

But if we are oing to have one:
```suggestion
     GenType type() const;
```

peterbarker commented on libraries/AP_Generator/AP_Generator_Backend.h:
```diff
@@ -0,0 +1,71 @@
+    // Constructor
+    AP_Generator_Backend(AP_Generator& frontend);
+
+    // Declare a virtual destructor incase Generator drivers want to override
```
```suggestion
    // Declare a virtual destructor in case Generator drivers want to override
```

peterbarker commented on libraries/AP_Generator/AP_Generator_Backend.h:
```diff
@@ -0,0 +1,71 @@
+    // Declare a virtual destructor incase Generator drivers want to override
+    virtual ~AP_Generator_Backend() {};
+
+    // Must all backends must use these functions
```
```suggestion
```

peterbarker commented on libraries/AP_Generator/AP_Generator_Backend.h:
```diff
@@ -0,0 +1,71 @@
+    int32_t _crit_fs_bitmask;
+
+    // Battery monitor readings to return
+    struct OutputReadings {
```
Unused?

Gone4Dirt replied:
_output is used

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    if (_uart != nullptr) {
+        _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_Generator, 0));
+    } else {
+        gcs().send_text(MAV_SEVERITY_INFO, "No serial port found");
```
More information needed here - it will be a context-free message to the user ATM.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
```
`const`

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    int16_t nbytes = _uart->available();
```
That's not the return type from `available()`.

The result from available() should be clamped to, say, 1024 bytes to make sure we don't spend too much time parsing here (it's a precaution we take elsewhere)

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+   // Read any available data
+    int16_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        char c = _uart->read();
```
That's not the return type from read().

-1 must be handled, even if the current code won't produce it with the structure you have here.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    int16_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        char c = _uart->read();
+        if (decode(c)) {
```
Invert this check and `continue;`

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+        char c = _uart->read();
+        if (decode(c)) {
+            // Successfully decoded a new reading
+            if (_type == AP_Generator::GenType::IE_650_800) {
```
Not reviewing this until we have >1 backend type...

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    if (_state.err_code != _state.last_err_code) {
+        if (_type == AP_Generator::GenType::IE_650_800) {
+            // Convert the error code to the failsafe bitmask for IE_650_800
+            for (const struct Lookup_FS_IE6 entry : lookup_fs_ie6) {
```
```suggestion
            for (const auto& entry : lookup_fs_ie6) {
```

Similarly below

Gone4Dirt replied:
Why do we want to auto here if we can easily identify the type?

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+// Check for arming
+bool AP_Generator_IE_FuelCell::arming_checks(char *failmsg, uint8_t failmsg_len)
+{
+    char message[45];
```
Why does this temporary variable exist?

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+
+    // Refuse arming if not healthy
+    if (!healthy()) {
+        strcpy(message, "Driver not healthy");
+        strncpy(failmsg, message, failmsg_len);
```
```suggestion
        strncpy(failmsg, "Driver not healthy", failmsg_len);
```

similarly elsewhere

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    }
+
+    // Check for error codes
+    if (check_err_code(message, true)) {
```
Pass failmsg and failmsg_len in instead.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+}
+
+// Lookup table for running state.  State code is the same for all IE units.
+const AP_Generator_IE_FuelCell::Lookup_State AP_Generator_IE_FuelCell::lookup_state[] = {
```
```suggestion
static const AP_Generator_IE_FuelCell::Lookup_State AP_Generator_IE_FuelCell::lookup_state[] = {
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,441 @@
+    _logging.last_time_ms = _state.last_time_ms;
+
+    AP::logger().Write(
+        "GEN",
```
Too generic.  Perhaps `IEGN`

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define TERM_BUFFER 12          // Max length of term we expect
```
This doesn't belong in the library's interface.  Just inline it.

Similarly elsewhere

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+#if GENERATOR_ENABLED
+
+#include <AP_Param/AP_Param.h>
+#include <AP_SerialManager/AP_SerialManager.h>
```
Don't need this include here.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+        TANK_LOW_PRESSURE       = 0x20000,//(1 << 17),
+    };
+
+    struct Lookup_FS_IE6 {
```
Don't need this in the header.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+        bool                    healthy;               // Is the driver working
+    } _state;
+
+    // Data that is received over telem and is only being logged
```
Unused

Gone4Dirt replied:
_logging is used.  This is now moved to IE_2400 backend only as this is the only unit that logs those values.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+        uint32_t                last_time_ms;          // Time we last wrote to log
+    } _logging;
+
+    // Temporary state params
```
Would be nice to put these into a structure, something like `struct parsed_value` or similar.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,178 @@
+    uint8_t _temp_state;
+    uint32_t _temp_err_code;
+
+    // Decodeing vars
```
Spolling error

peterbarker reviewed (changes requested):
Preliminary review only.

Gone4Dirt commented:
Made corrections and split out into backend.  Still not passing Richenpower autotest and still need to hide from 1mb boards so still not ready for merge yet.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator.h:
```diff
@@ -0,0 +1,110 @@
+        rpm,
+    };
+
+    // Helpers to retrieve measurements
```
The values being returned are in private so that they can't be modified from outside the library.  Is not the correct way of doing this?  What did you have in mind?

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,60 @@
+
+
+private:
+    // Process characters recieved and build sentence for IE 2.4kW
```
Is this really what it does?  "build sentence"s?

Perhaps "extract terms" might be better?

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,60 @@
+
+    static const Lookup_FS_IE24 lookup_fs_ie24[];
+
+    // Data that is received over telem and is only being logged
```
Not sure you need this structure here.  Also, last_time_ms is not received over telemetry :-)

Gone4Dirt replied:
So, I am a little confused as to when it is appropriate to add a structure in AP.  I appreciate that only one instance is ever created here, but it acts a a good collector for the variables to easily see where they are used.  You asked me to add a structure for a similar purpose for 'parsed variables' so I am now a little confused.  Can you you give me some guidance please.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.cpp:
```diff
@@ -0,0 +1,163 @@
+    const uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    uint32_t nbytes = MIN(_uart->available(),30u);
```
This loop should be moved to the base class, with the subclasses being called to handle the decoded sentences.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.cpp:
```diff
@@ -0,0 +1,163 @@
+{
+    // Null terminate and move onto the next term
+    _term[_term_offset] = 0;
+    _term_offset = 0;
```
Doesn't need changing, but I would have used zero-indexing on _term_number and moved this incrementing and resetting of _term_offset to the bottom of decode_latest_term.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.h:
```diff
@@ -0,0 +1,59 @@
+    // IE 650/800 W error codes and corresponding lookup table
+    // Note: The error codes for this unit are defined as a bitmask
+    enum class ERR_CODE_IE6 : uint32_t {
+        STACK_OT1               = 0x80000000,//(1 << 31),
```
```suggestion
        STACK_OT1               = (1U<<31),
```

etc

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    if (_uart != nullptr) {
+        _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_Generator, 0));
+    } else {
+        gcs().send_text(MAV_SEVERITY_INFO, "Generator: No serial port found");
```
Switch around and do an early-return

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,148 @@
+{
+    // If fuel cell state has changed send gcs message
+    if (_state.state != _state.last_state) {
+        for (const struct Lookup_State entry : lookup_state) {
```
```suggestion
        for (const struct Lookup_State &entry : lookup_state) {
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,148 @@
+    if (_state.state != _state.last_state) {
+        for (const struct Lookup_State entry : lookup_state) {
+            if (_state.state == entry.option) {
+                gcs().send_text(MAV_SEVERITY_ALERT, "Fuel cell %s", entry.msg_txt);
```
```suggestion
                gcs().send_text(MAV_SEVERITY_ALERT, "Fuel cell: %s", entry.msg_txt);
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    // Check error codes
+    char msg_txt[32];
+    if (check_err_code(msg_txt, 32, false)) {
```
```suggestion
    if (check_err_code(msg_txt, sizeof(msg_text), false)) {
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+#if GENERATOR_ENABLED
+
+#include <AP_Param/AP_Param.h>
+#include <AP_SerialManager/AP_SerialManager.h>
```
This doesn't need to be in the header.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+
+#include "AP_Generator_Backend.h"
+
+#ifndef GENERATOR_ENABLED
```
This shouldn't need to be redefine here.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define TERM_BUFFER            12  // Max length of term we expect
```
This define shouldn't be here.  I'd suggest just putting it inline where it is being used.

Out here it pollutes everybody who includes this file, directly and indirectly.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+#include <GCS_MAVLink/GCS.h>
+
+#define TERM_BUFFER            12  // Max length of term we expect
+#define HEALTHY_TIMEOUT_MS   5000  // Time for driver to be marked un-healthy
```
Shouldn't be out here either.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+    AP_HAL::UARTDriver *_uart = nullptr;
+
+    // IE fuel cell running states and corresponding lookup table
+    enum class FUEL_CELL_STATE {
```
Given this is enum-class within `AP_Generator_IE_FuelCell`, this can just be `State`.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+    static const Lookup_State lookup_state[];
+
+    // The fuel cell state structure
+    struct FuelCellState {
```
Don't need a separate structure to hold these now.

peterbarker reviewed (changes requested):
Several of my comments for the 2.4kWh backend apply to the other.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator.h:
```diff
@@ -0,0 +1,110 @@
+
+class AP_Generator
+{
+    friend class AP_Generator_Backend;
```
My understanding is that the back end needs to be a friend class to access the protected variables in the front end.  The fuel cell base class needs to be a friend to access the parameters defined in the front end.

Gone4Dirt replied:
Sorry, still not getting this.  The backends don't inherit from the front end so won't have access to protected.  This approach is implemented in other libraries (e.g. AP_BattMonitor, AP_RPM).  What am I missing here?

Gone4Dirt commented on libraries/AP_BattMonitor/AP_BattMonitor_Generator.cpp:
```diff
@@ -1,38 +1,99 @@
+// Read the fuel level.  Should be called at 10hz
+void AP_BattMonitor_Generator_Fuel::read()
+{
+    common_checks();
```
Changed name as requested.  I don't understand how generator can be nullptr after this line.  The nullptr check is done in 'generator_healthy()' which is always called first in read().

Gone4Dirt replied:
This has now changed such that there was little commonality between the two sub-classes.  I have now gotten rid of the base class.

Gone4Dirt commented on libraries/AP_BattMonitor/AP_BattMonitor_Generator.cpp:
```diff
@@ -1,38 +1,99 @@
+    Electrical class
+*/
+// Read the electrical measurements from the generator
+void AP_BattMonitor_Generator_Elec::read()
```
Disagree here.  Generator_Elec refers to the electrical measurements made on the generator.  Generator_Fuel refers to the fuel level measurements made on the generator.  I feel that is a sufficient distinction.

Gone4Dirt commented on libraries/AP_BattMonitor/AP_BattMonitor_Generator.cpp:
```diff
@@ -1,38 +1,99 @@
+    // Only check for fuel cell failsafes on the electical moniter
+    // no point in having the same failsafe on two battery monitors
+    if (generator != nullptr) {
+            failsafe = generator->update_failsafes();
```
I agree with what you are saying, however, I have adopted the naming convention of the library.  update_failsafes in Batt_Monitor returns a value so I have adopted the same approach for constancy in this instance.

Gone4Dirt commented on libraries/AP_BattMonitor/AP_BattMonitor_Generator.h:
```diff
@@ -2,21 +2,64 @@
+    // Empty init to satisfy pure virtual function in AP_BattMonitor_Backend
+    void init(void) override {};
+
+    // Always return true, if the generator doesnt actually have current info it will just be zero value
```
I originally had as you are suggesting.  However, IMHO there is little point in Batt_Monitor asking generator if it has current just to set it to zero.  To ask a specific generator if it has current it generates a lot of additional functions to go through the code path: e.g. AP_BattMonitor_Generator->AP_Generator.cpp->AP_Generator_Backend.cpp->AP_Generator_IE_FuelCell.cpp->AP_Generator_IE_2400.cpp.  So there is less additional code to just set it to has_current = true and the value of current is left as zero.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator.cpp:
```diff
@@ -0,0 +1,185 @@
+
+    // @Param: FS_LOW
+    // @DisplayName: Generator failsafe trigger Low action
+    // @Description: Failsafe bitmask that triggers battery failsafe low action. Bits are shared for different generator types.  Please see AP wiki page for table of failsafe bits.
```
Added more to the metadata to explain.  The bitmask is shared across all generators.  IE_650_800 has 24 failsafe bits so it seems a waste to add more parameters so the metadata can match the specific bits to the failsafes of a particular generator.  So I figured the best compromise is to give users a table to look up in the Wiki.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,217 @@
+    const uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    uint32_t nbytes = MIN(_uart->available(),30u);
```
Yeah, thankfully IE were very prescriptive in the data types and size being sent.  It will take the largest string length from a 'valid sentence' from either of the fuel cell units.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,217 @@
+            break;
+
+        default:
+            // If we get any other term number something has gone wrong
```
Sorry, don't understand what you are looking for here.  Can you elaborate please.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,217 @@
+}
+
+// Lookup table for IE 2.4 kW fuel cell failsafes.
+const AP_Generator_IE_2400::Lookup_FS_IE24 AP_Generator_IE_2400::lookup_fs_ie24[] = {
```
Sorry, not sure what you are asking for here.  Can you elaborate please?

Gone4Dirt commented:
Moved more of generator behind HAL_Minimise.  When building for a 1mb board it saves 296 b of flash

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+private:
+
+    // Assigns the unit specific measurements once a valid sentence is obtained
+    void assign_measurments(const uint32_t now) override;
```
```suggestion
    void assign_measurements(const uint32_t now) override;
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+    // Assigns the unit specific measurements once a valid sentence is obtained
+    void assign_measurments(const uint32_t now) override;
+
+    // Process characters recieved and extract terms for IE 2.4kW
```
```suggestion
    // Process characters received and extract terms for IE 2.4kW
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+    // Convert error code from fuel cell to AP failsafe bitmask
+    void set_failsafe_mask(uint32_t err_in);
+
+    // Check if we have recieved an error code and populate message with error code
```
```suggestion
    // Check if we have received an error code and populate message with error code
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+
+    static const Lookup_FS_IE24 lookup_fs_ie24[];
+
+    // Data that is recieved over telem and is only being logged
```
```suggestion
    // Data that is received over telem and is only being logged
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+    struct Log_Only_Data {
+        int16_t    pwr_out;       // Output power (Watts)
+        uint16_t   spm_pwr;       // SPM power draw (Watts)
+        uint32_t   last_time_ms;  // Time we last wrote to log
```
This is neither received over telemetry nor logged.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,188 @@
+}
+
+// Check error codes and populate message with error code
+bool AP_Generator_IE_2400::check_for_err_code(char* msg_txt, uint8_t msg_len, bool changed_check)
```
This might be neater factored into two methods and doing without changed_check.

So one method just returns false if there are not errors, the second wraps that one with the additional "has it changed" check before returning zero.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,188 @@
+    return true;
+}
+
+// log generator status to the onboard log
```
Why isn't this being done for the other unit too?  Why isn't it in the common base class?

Gone4Dirt replied:
The other unit doesn't have the measurements that are being logged here. This unit logs these measurements as they cannot be obtained from the Batt Monitor log. All of the measurements on the other fuel cell unit can be obtained from Batt_Monitor log so would just be a duplication. The exception being the error code on the 650/800W unit, but I figure that the error code is already being printed to gcs as a hex code (which is more meaningful to the users). Having a log_write() in the 650/800W just to log the decimal value seems unneeded.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.h:
```diff
@@ -0,0 +1,61 @@
+private:
+
+    // Assigns the unit specific measurements once a valid sentence is obtained
+    void assign_measurments(const uint32_t now) override;
```
```suggestion
    void assign_measurements(const uint32_t now) override;
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.h:
```diff
@@ -0,0 +1,61 @@
+    // Assigns the unit specific measurements once a valid sentence is obtained
+    void assign_measurments(const uint32_t now) override;
+
+    // Process characters recieved and extract terms for IE 650/800 W
```
```suggestion
    // Process characters received and extract terms for IE 650/800 W
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+
+public:
+    // Constructor
+    AP_Generator_IE_FuelCell(AP_Generator &frontend);
```
```suggestion
    using AP_Generator_Backend::AP_Generator_Backend;
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp:
```diff
@@ -0,0 +1,184 @@
+   // Read any available data
+    uint32_t nbytes = MIN(_uart->available(),30u);
+    while (nbytes-- > 0) {
+        int16_t c = _uart->read();
```
```suggestion
        const int16_t c = _uart->read();
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+    // Pointer to serial uart
+    AP_HAL::UARTDriver *_uart = nullptr;
+
+    // IE fuel cell running states and corresponding lookup table
```
```suggestion
    // IE fuel cell running states and corresponding to-string lookup table
```

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_FuelCell.h:
```diff
@@ -0,0 +1,108 @@
+        uint32_t      failsafe;       // The failsafe state
+        uint32_t      err_code;       // The error code from fuel cell
+        uint32_t      last_err_code;  // The previous error code from fuel cell
+        enum State    state;          // The PSU state
+        enum State    last_state;     // The previous PSU state
```
```suggestion
        State    state;          // The PSU state
        State    last_state;     // The previous PSU state
```

peterbarker reviewed (changes requested):
Looking much nicer here now.

peterbarker reviewed (changes requested):
I've run this PR on my RichenPower generator here and it works as expected.

peterbarker commented:
I've rebased and fixed up the ~~crappy little~~ low-fidelity simulator I put together for the IE24 and pushed the branch up here: https://github.com/peterbarker/ardupilot/tree/xwip/intelligent_energy_fuel_cell_simulator

I'll merge that after this goes in, or you could cherry-pick it on top of this PR if you felt like it, @Gone4Dirt 

I also added a trivial autotest that you can't arm if you're battery-only to that branch.

Gone4Dirt commented on libraries/AP_DAL/AP_DAL.h:
```diff
@@ -11,6 +11,7 @@
 
 #include "LogStructure.h"
 
+#include <stdio.h>
```
It was failing CI.  I managed to reproduce locally.  Fails for linux boards (I used navio) and ./waf all .  Doesn't fail for a specific vehicle build.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator.h:
```diff
@@ -0,0 +1,106 @@
+    float get_batt_consumed(void) const { return _consumed_mah; }
+    uint16_t get_rpm(void) const { return _rpm; }
+
+    // Helpers to see if backend has a measurement
+    bool has_current() const { return _has_current; }
+    bool has_consumed_energy() const { return _has_consumed_energy; }
+    bool has_fuel_remaining() const { return _has_fuel_remaining; }
```
I was torn.  Doing it this way required less functions to get the values from the backend but meant I had add more friend classes.  I figured having the backends as a friend class wasn't the worst thing in the world.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,206 @@
+// Check error codes and populate message with error code
+bool AP_Generator_IE_2400::check_for_err_code(char* msg_txt, uint8_t msg_len) const
+{
+    // Check if we have reveiced an error code
```
If this PR has taught me anything its that I can't spell received :-)

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_2400.h:
```diff
@@ -0,0 +1,56 @@
+
+    // These measurements are only available on this unit
+    int16_t _pwr_out;  // Output power (Watts)
+    uint16_t _spm_pwr; // SPM power draw (Watts)
```
Stack Power Module (SPM).  Have added to comment.

Gone4Dirt commented on libraries/AP_Generator/AP_Generator_IE_650_800.cpp:
```diff
@@ -0,0 +1,161 @@
+    // Invert bat remaining percent to match AP_BattMonitor convention
+    _consumed_mah = 100.0f - _parsed.battery_pct;
+
+    // This unit does not report voltage, always report 1 volt
+    _voltage = 1;
```
Yes, that would be nice.  Happy to implement that but would prefer to do it as a separate PR to not hold this one up.

magicrub commented on libraries/AP_ADSB/AP_ADSB.cpp:
```diff
@@ -643,7 +642,7 @@ void AP_ADSB::set_callsign(const char* str, const bool append_icao)
 
     if (append_icao) {
-        snprintf(&out_state.cfg.callsign[4], 5, "%04X", unsigned(out_state.cfg.ICAO_id % 0x10000));
+        hal.util->snprintf(&out_state.cfg.callsign[4], 5, "%04X", unsigned(out_state.cfg.ICAO_id % 0x10000));
```
What's this change for?

Gone4Dirt replied:
One of my includes of stdio.h has resulted in a build error on Linux boards because of an include order dependency caused by the the snprintf in AP_ADSB.cpp no longer being able to access stdio.h.  Having spoken to @peterbarker  his recommendation was that the snprintf in AP_ADSB should be be using the same function from the hal.  I don't understand the specifics as to why though.  So I can either make this change or include the stdio.h explicitly AP_ADSB.cpp.  There is no functional change.

magicrub replied:
Well, of simply including stdio.h solves it then that seems the best solution. You added stdio to AP_DAL, might as well do it here too.

rmackay9 replied:
I don't think there's anything wrong with changing this to use hal-util->snprintf() so I think we should avoid the back-and-forth of asking @Gone4Dirt to change it again.  From a quick search through the code we seem to use either.  If we were going to consolidate on one of the two I suspect we would consolidate on the hal_util version.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.cpp:
```diff
@@ -0,0 +1,161 @@
+bool AP_Generator_IE_650_800::check_for_err_code(char* msg_txt, uint8_t msg_len) const
+{
+    // Mask for all valid error codes
+    const uint32_t ALL_ERROR_CODES = 4294967288;
```
Still waiting on a response here.

Gone4Dirt replied:
Done as requested

Gone4Dirt commented:
Thanks @peterbarker for the autotest.  I have cherry-picked it to this PR. I have made some modifications.  I changed some of the data fields on the telem message to match the spec that was provided to me by IE.  I have also added an error parameter and an autotest check to make sure that the failsafe behaviour doesn't rot.

peterbarker commented on libraries/AP_Generator/AP_Generator_IE_650_800.h:
```diff
@@ -0,0 +1,98 @@
+
+    // Assign error codes to critical FS mask
+    static const uint32_t fs_crit_mask = ERROR_STACK_OT1_ALRT // (0x80000000) Stack 1 over temperature alert (>58 degC)
+                                + ERROR_STACK_OT2_ALRT        // (0x40000000) Stack 2 over temperature alert (>58 degC)
```
bitmasks are traditionally assembled with `|` rather than `+`.

Not sure how I missed that until now.

I don't see this as a blocker.

Gone4Dirt replied:
fixed

tridge commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,206 @@
+        return false;
+    }
+
+    snprintf(msg_txt, msg_len, "Fuel cell err code <%u>", (unsigned)_err_code);
```
hal.util->snprintf

Gone4Dirt replied:
Done

tridge commented on libraries/AP_Generator/AP_Generator_IE_2400.cpp:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    // Calculate consumed current
+    const float AMS_TO_MAH = 0.000277777778f;  // Amp.millisec to milliAmp.hour = 1/1E3(ms->s) * 1/3600(s->hr) * 1000(A->mA)
```
add to definitions.h

Gone4Dirt replied:
Done

tridge commented on libraries/AP_Generator/AP_Generator_IE_650_800.cpp:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    // Error codes are converted to hex to make it easier to compare to user manual for these units
+    snprintf(msg_txt, msg_len, "Fuel cell err code <0x%x>", (unsigned)_err_code);
```
hal.util->

tridge commented on libraries/SITL/SIM_IntelligentEnergy24.cpp:
```diff
@@ -0,0 +1,127 @@
+    // @Description: Allows you to enable (1) or disable (0) the FuelCell simulator
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO("ENABLE", 0, IntelligentEnergy24, enabled, 0),
```
don't use index 0 for parm tables

Gone4Dirt replied:
Done

tridge commented on libraries/SITL/SIM_IntelligentEnergy24.cpp:
```diff
@@ -0,0 +1,127 @@
+    last_sent_ms = now;
+
+    char message[128];
+    snprintf(message, ARRAY_SIZE(message), "<%i,%.1f,%i,%u,%i,%u,%u>\n",
```
hal.util->

tridge commented on libraries/AP_Math/definitions.h:
```diff
@@ -106,4 +106,6 @@ static const double WGS84_E = (sqrt(2 * WGS84_F - WGS84_F * WGS84_F));
-
+// Convert amps milliseconds to milliamp hours
+// Amp.millisec to milliAmp.hour = 1/1E3(ms->s) * 1/3600(s->hr) * 1000(A->mA)
+const float AMS_TO_MAH = 0.000277777778f;
```
should be a define

tridge commented:
once defs are fixed and passes CI can be merged

peterbarker commented:
Merged, thanks!
